main/src/ca/uwaterloo/flix/util/Validation.scala

```diff
@@ -214,54 +262,38 @@ object Validation {
     */
   def mapN[T1, T2, U, E](t1: Validation[T1, E], t2: Validation[T2, E])
                         (f: (T1, T2) => U): Validation[U, E] =
-    (t1, t2) match {
-      case (Success(v1), Success(v2)) => Success(f(v1, v2))
-      case _ => Failure(t1.errors #::: t2.errors)
-    }
+    ap(mapN(t1)(curry(f)))(t2)
```

Comment:
I understand you want to do this to make the code pretty, but it incurs a performance cost.

We have typically written Validation to be stupid, but fast.

Here I would *suspect* you would introduce a lot of additional overhead.

You can try with --Xbenchmark-throughput to see if that is the case.

Reply:
Alright, I'll try that

Reply:
But keep in mind that for `mapN` that takes 9 parameters, if we were to keep it naive without the performance cost then we would have $2^9 + 1= 513$ distinct combinations i.e. pattern matches (assuming we use a wildcard pattern for the failure cases).

Reply:
I think we should benchmark the two new solutions versus the current implementation.